Updating docs (#753)
grafana/azure-data-explorer-datasource: README.md

   ```

3. Open the browser at: <http://localhost:3000> or <http://your-domain-name:3000>
4. Login in with username: `admin` and password: `admin`
5. To make sure the plugin was installed, check the list of installed datasources. Click the Plugins item in the main menu. Both core datasources and installed datasources will appear.

This ia an alternative command if you want to run Grafana on a different port than the default 3000 port:

```bash
docker run -d --name=grafana -p 8081:8081 -e "GF_SERVER_HTTP_PORT=8081" -e "GF_INSTALL_PLUGINS=grafana-azure-data-explorer-datasource" grafana/grafana:master
```

It is recommended that you use a volume to save the Grafana data in. Otherwise if you remove the docker container, you will lose all your Grafana data (dashboards, users etc.). You can create a volume with the [Docker Volume Driver for Azure File Storage](https://github.com/Azure/azurefile-dockervolumedriver).

#### Docker latest

Alternatively, you can use the tag `azure-data-explorer-datasource-latest` to pull down a Docker image for whatever is current in `main`.

> `docker run --rm -p 3000:3000 grafana/plugin-builds:azure-data-explorer-datasource-latest`

Reviewer: I don't think the grafana docs touch on Docker, do we want to move this info to a different file called `DOCKER.md`?

Reviewer 2: That information is [here](https://grafana.com/docs/grafana/latest/setup-grafana/installation/docker/#install-plugins-in-the-docker-container) so I think it's fine to remove. The only difference I'd say is the ability to pull down whatever's in current main for ADX, that's something I don't think is generally available for all plugins?

Reviewer: Ah I knew I was missing something Docker related somewhere, thank you! Yes, the only difference is we have a "pointer container"  to `main`. Maybe we leave that in? Or should we move it out?

Author: I think we should move this out. I don't think someone trying to set up Grafana with the ADX plugin will likely find this information. Also, would a user ever want to use the main instead of the latest release? I feel like this is a note for developers.

enforce_trusted_endpoints = true
```

## Configure the Azure Data Explorer data source

> To configure ADX for using this data source, create an Azure Active Directory (AAD) Application and AAD Service Principle, log into the Azure Data Explorer WebExplorer and connect the AAD Application to an Azure Data Explorer database user, and use the AAD Application to configure the data source connection in Grafana.

Reviewer 3: I would actually suggest formatting each action as its own step, and I really think users would benefit instruction on how to actually accomplish these actions instead of just being told to do them unless it's super obvious. 

```suggestion
To configure ADX for using this data source:
1. Create an Azure Active Directory (AAD) Application and AAD Service Principle.
1. In the Azure Data Explorer WebExplorer and connect the AAD Application to an Azure Data Explorer database user.
1. Use the AAD Application to configure the data source connection in Grafana.
```

> ***Optionally** To use the dropdown cluster select when creating queries, add reader access for Azure Resource Graph queries, and add reader access to each desired cluster.

Reviewer 3: ```suggestion
1. (Optional) To use the dropdown cluster select when creating queries, add reader access for Azure Resource Graph queries, and add reader access to each desired cluster.
```

enforce_trusted_endpoints = true
```

## Configure the Azure Data Explorer data source

To configure ADX for using this data source:

> 1. Create an Azure Active Directory (AAD) Application and AAD Service Principle.

Reviewer 2: ```suggestion
1. Create an Azure Active Directory (AAD) Application and AAD Service Principal.
```

> 1. In the Azure Data Explorer WebExplorer and connect the AAD Application to an Azure Data Explorer database user.

Reviewer 2: ```suggestion
1. In the Azure Data Explorer WebExplorer, connect the AAD Application to an Azure Data Explorer database user.
```

1. Use the AAD Application to configure the data source connection in Grafana.

> 1. (Optional) To use the dropdown cluster select when creating queries, add reader access for Azure Resource Graph queries, and add reader access to each desired cluster.

Reviewer 2: I'm fairly certain `Reader` access isn't needed for each cluster in addition to each subscription? It's either or right?

Author: We have reader access to the grafana cluster we do not own and cannot see it in the cluster select.

Reviewer 2: In that case we _should_ be able to remove the `and add reader access to each desired cluster.` line as it's the subscription reader access that's relevant

### Creating an Azure Active Directory Service Principle

> For detailed instructions on how to set up a Microsoft Entra application and service principal that can access resources, please follow this guide from Microsoft [Create a Microsoft Entra application and service principal that can access resources](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal)

Reviewer 2: ```suggestion
For detailed instructions on how to set up a Microsoft Entra application and service principal that can access resources, please follow this guide from Microsoft: [Create a Microsoft Entra application and service principal that can access resources](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal)
```

| Cache max age | By default, the cache is disabled. If you want to enable the query caching please specify a max timespan for the cache to live. |
| Data consistency | Query consistency controls how queries and updates are synchronized. Defaults to Strong. For more information, refer to [Query consistency](https://learn.microsoft.com/en-us/azure/data-explorer/kusto/concepts/queryconsistency) | 
| Default editor mode | This setting dictates which mode the editor will open in. Defaults to Visual. |
| Default database | The default database will be used if no database is selected. A default cluster is required to select a default database. To load default database options, you must save the data source with a valid Azure connection. |
| Use managed schema | If enabled, tables, functions, and materialized views are mapped to user-friendly names. |
| Send username header to host | With this feature enabled, Grafana will pass the logged in user's username in the `x-ms-user-id` header and in the `x-ms-client-request-id` header when sending requests to ADX. It can be useful when tracking needs to be done in ADX. |

### Configuring On-Behalf-Of authentication (Beta)

> _This feature is in Beta and subject to breaking changes_

Reviewer 3: ```suggestion
{{%/* admonition type="caution" */%}}
This feature is in Beta and subject to breaking changes
{{%/* /admonition */%}}
```

Reviewer 3: just a heads up that if you're building your branch locally, the admonition shortcode I added here might not render for you, but it should render live when you publish/merge

Author: oooo this is really cool!

> For information about setting up and using the OBO flow: [on-behalf-of documentation](https://github.com/grafana/azure-data-explorer-datasource/blob/main/doc/on-behalf-of.md)

Reviewer 3: ```suggestion
For information about setting up and using the OBO flow, refer to [on-behalf-of documentation](https://github.com/grafana/azure-data-explorer-datasource/blob/main/doc/on-behalf-of.md)
```

## Query the data source

Before querying the data source, select the query header options cluster, database, and format. You can create queries using the query builder, KQL, or OpenAI.

### Query header

#### Cluster

Select a cluster to query. If a default cluster was set in the data source settings, then it will auto-populate the cluster select. If there are no clusters to choose from, refer to [Configure the Azure Data Explorer data source](#configure-the-azure-data-explorer-data-source)

#### Database

Select a database to query. If a default database was set in the data source settings, it will auto-populate the database selection. 

#### Format as

Queries can be formatted as _Table_, _Time Series_, _Trace_, or _ADX time series_ data using the **Format as** dropdown select. 

> ##### Table

Reviewer 3: I suggest avoiding H5s whenever possible (it's hard for human eyes to distinguish them below the H4 level, and I don't think they show up in navigation on the left or right TOCs); I wonder if it would make sense to either add these queries as bullet points or in some other way, but since I'm not a subject matter expert, I'll not push the issue. 
I really like the way the Splunk data source doc is organized...but I don't know if that can be applied here. https://grafana.com/docs/plugins/grafana-splunk-datasource/latest/

Reviewer 3: Just realized that this is a different format since it's a single README file.

Author: Thank you! I had a feeling I was doing something wrong with H5s. I'll reformat to use bullet points similar to Splunk's query editor bullet points.

| Group by | (Optional) Add group bys for the selected columns. For time group bys select a time range bucket. |
| Timeshift | (Optional) **DO NOT MERGE!** why does this exist? |

Columns of the `dynamic` type are supported within the query builder. This encompasses arrays, JSON objects, and nested objects within arrays. A limitation is only the first 50,000 rows are queried for data, so only properties contained within the first 50,000 rows will be listed as options in the builder selectors. Additional values can be manually written in the different selectors if they don't appear by default. Also, due to the fact that these queries make use of `mv-expand`, they may become resource intensive.

Refer to the documentation below for further details on handling dynamic columns appropriately via the KQL editor.

[Kusto Data Types](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/scalar-data-types/) - Documentation on data types supported by Kusto.

[Dynamic Data Type](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/scalar-data-types/dynamic) - Detailed documentation on the dynamic data type.

### Query with Kusto Query Language (KQL)

> Queries are written in the new Kusto Query Language; for more information, refer to [Kusto Query Language (KQL) overview](https://docs.microsoft.com/en-us/azure/kusto/query/).

Reviewer 2: ```suggestion
Queries are written in Kusto Query Language; for more information, refer to [Kusto Query Language (KQL) overview](https://docs.microsoft.com/en-us/azure/kusto/query/).
```

  | extend series_decompose_forecast(AvgHatInventory, 30) | project-away *residual, *baseline, *seasonal
```

### Query Builder

| Field | Description |
| -- | -- |
| Table | Select a table. |
| Columns | Select a subset of columns for faster results. Time series requires both time and number values; other columns are rendered as dimensions. For more information about dimensions, refer to [Time series dimensions](https://grafana.com/docs/grafana/latest/fundamentals/timeseries-dimensions/). |
| Filters | (Optional) Add filters for the selected columns. Values for filters will be restricted to the column's data type. |
| Aggregate | (Optional) Add aggregations for the selected columns. Select an aggregation type from the dropdown and select a column to aggregate on.  |
| Group by | (Optional) Add group bys for the selected columns. For time group bys select a time range bucket. |

> | Timeshift | (Optional) **DO NOT MERGE!** why does this exist? |

Reviewer 3: is the Timeshift entry intentional?

Author: @aangelisc @asimpson Do y'all know?

Reviewer 2: This was added in #142 and I understand what it does based on testing but tbh looking at [this](https://github.com/grafana/grafana/issues/24784#issuecomment-704160213) comment we may be able to remove this field entirely in future (once we migrate queries). 

In order to update the `README` appropriately I'd say something along the lines of `Shift the time ranges generated from Grafana macros by a predetermined duration` or similar.

Refer to the documentation below for further details on handling dynamic columns appropriately via the KQL editor.

[Kusto Data Types](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/scalar-data-types/) - Documentation on data types supported by Kusto.

[Dynamic Data Type](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/scalar-data-types/dynamic) - Detailed documentation on the dynamic data type.

### Query with Kusto Query Language (KQL)

Queries are written in the new Kusto Query Language; for more information, refer to [Kusto Query Language (KQL) overview](https://docs.microsoft.com/en-us/azure/kusto/query/).

### OpenAI query generator

> _You need to enable the LLM plugin to use this feature._

Reviewer 3: ```suggestion
{{%/* admonition type="note" */%}}
You must enable the LLM plugin to use this feature.
{{%/* /admonition */%}}

```

## Templating with Variables

Instead of hard-coding things like server, application, and sensor name in your metric queries, you can use variables in their place. Variables are shown as dropdown select boxes at the top of the dashboard. These dropdowns make it easy to change the data being displayed in your dashboard.

> Create the variable in the dashboard settings. Usually, you will need to write a query in the Kusto Query Language to get a list of values for the dropdown. However, having a list of hard-coded values is also possible.

Reviewer 2: ```suggestion
Create the variable in the dashboard settings. Usually, you will need to write a query in KQL to get a list of values for the dropdown. However, having a list of hard-coded values is also possible.
```
